### ISSUE TITLE
The `disallowed_types` lint also lints when the type comes from a macro expansion in a different crate

### ISSUE THREAD

**Issue by TheNeikos** [rust-lang/rust-clippy#17718]
### Summary

Using the `disallowed_types` can cause problems when the type in question is generated by a proc-macro in a crate that is out of control of the user. Since the type can be used an arbitrary position, there is no way to attach an `allow` or similar block to it, and as such means that one has to allow the whole module, or drop the configured disallowed type or external crate, all of which are not great.

### Lint Name

disallowed_types

### Reproducer

I tried this code:

With my configuration disallowing `std::path::PathBuf`

```rust
#[derive(displaydoc::Display)]
enum Test {
  /// A foo error
  Foo
}
```

I saw this happen:

```
warning: use of a disallowed type `std::path::PathBuf`
```

I expected to see this happen: [nothing]

### Version

```text
rustc 1.97.1 (8bab26f4f 2026-07-14)
binary: rustc
commit-hash: 8bab26f4f68e0e26f0bb7960be334d5b520ea452
commit-date: 2026-07-14
host: x86_64-unknown-linux-gnu
release: 1.97.1
LLVM version: 22.1.6
```
